core/src/main/java/com/linecorp/armeria/server/ExceptionReportingServerErrorHandler.java

```diff
@@ -0,0 +1,152 @@
+ */
+
+package com.linecorp.armeria.server;
+
+import java.time.Duration;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.LongAdder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.TextFormatter;
+import com.linecorp.armeria.server.logging.LoggingService;
+
+import io.micrometer.core.instrument.Tags;
+
+/**
+ * A {@link ServerErrorHandler} that wraps another {@link ServerErrorHandler}
+ * to periodically report the exceptions that were not logged by decorators such as
+ * {@link LoggingService}.
+ */
+final class ExceptionReportingServerErrorHandler implements ServerErrorHandler, ServerListener {
+
+    private static final Logger logger = LoggerFactory.getLogger(ExceptionReportingServerErrorHandler.class);
+
+    @Nullable
+    private static ExceptionReportingServerErrorHandler exceptionReportingServerErrorHandler;
+    private final ServerErrorHandler delegate;
+    private final Duration duration;
+    private final LongAdder counter;
+    private long lastExceptionsCount;
+    @Nullable
+    private Throwable thrownException;
+    @Nullable
+    private ScheduledFuture<?> reportingTaskFuture;
+
+    private ExceptionReportingServerErrorHandler(ServerErrorHandler serverErrorHandler, Duration duration) {
+        assert !duration.isNegative() && !duration.isZero();
+        delegate = serverErrorHandler;
+        this.duration = duration;
+        counter = new LongAdder();
+    }
+
+    static ExceptionReportingServerErrorHandler of(ServerErrorHandler serverErrorHandler, Duration duration) {
+        if (exceptionReportingServerErrorHandler != null) {
+            return exceptionReportingServerErrorHandler;
+        }
+
+        exceptionReportingServerErrorHandler = new ExceptionReportingServerErrorHandler(serverErrorHandler,
+                                                                                        duration);
+        return exceptionReportingServerErrorHandler;
+    }
+
+    /**
+     * This method increments the {@link #counter} and stores unhandled exception that occurs. If unhandled
+     * exceptions are not previously scheduled for reporting, then they are scheduled to be reported.
+     */
+    @Nullable
+    @Override
+    public HttpResponse onServiceException(ServiceRequestContext ctx, Throwable cause) {
+        final boolean isExpectedException =
+                (cause instanceof HttpStatusException || cause instanceof HttpResponseException) &&
+                cause.getCause() != null;
+
+        if (ctx.shouldReportUnhandledExceptions() && !isExpectedException) {
+            if (reportingTaskFuture == null) {
```

Comment:
We can't guarantee what thread this will be called from, which can result in multiple reporting tasks.

What do you think of updating this atomically?


```suggestion
            if (reportingFutureInitializedUpdater.compareAndSet(this, false, true)) {
```

where the updater can be initialized like the following:
```
private static final AtomicReferenceFieldUpdater
        <ExceptionReportingServerErrorHandler, Boolean> reportingFutureInitializedUpdater =
        AtomicReferenceFieldUpdater.newUpdater(ExceptionReportingServerErrorHandler.class, Boolean.class,
                                               "reportingFutureInitialized");
```

Reply:
I think this appeared as a part of the singleton optimization.

Could we go back to the original idea of starting and stopping explicitly, following the life cycle of the `Server` it belongs to? Alternatively, we could make use of the `serverStarting` and `serverStopped` callbacks to schedule and unschedule as suggested by other reviewers.

Reply:
It is my suggestion that schedules the reporting task lazily only when an unhandled exception is found. This is because I believe most Armeria users already use `LoggingService` as a basic decorator. For those users, the reporting task won't be useful.

Reply:
Ah, I see. That's a good idea. We could start the scheduled task lazily and stop at `serverStopped`, then.